lifecycle api, types, tests & example

wayneparrott opened the pull request:
This PR implements the ROS2 Lifecycle Node (managed node) as described in the [ROS2 Managed Node design](https://design.ros2.org/articles/node_lifecycle.html) and implemented by the [rclcpp lifecycle implementation](https://github.com/ros2/rclcpp/tree/master/rclcpp_lifecycle).

### Key changes:

**index.js (updated)**
* added lifecyle import
* added rclnodejs.lifecycle property - simulate namespace
* added createLifecycleNode()
* refactored common logic for node creation for use in both createNode() and createLifecycleNode()

**index.ts (updated)**
* added TypeScript declarations for createLifecyleNode()

**lifecycle.js (new)**
* includes LifecycleNode, a subclass of ShadowNode, and lifecycle support classes and interfaces

**index.d.ts (new)**
* includes TypeScript declarations for LifecycleNode and lifecycle support classes and interfaces.

**rcl_lifecycle_bindings.hpp (new)**
**rcl_lifecycle_bindings.cpp (new)**
* contains native addon C functions for interfacing with rcl_lifecycle_state_machine data-structures and api.

**lifecycle_publisher.js (new)**
* implements LifecyclePublisher - based on the rclcpp LifecyclePublisher implementation

**lifecycle_publisher.d.ts (new)**
* TypeScript declarations for LifecyclePublisher.

**publisher.js (updated)**
* refactored common publisher creation functionality for reuse by LifecyclePublisher.

**test-lifecycle.js (new)**
* unit tests for LifecycleNode and related classes

**test-lifecycle-publisher.js (new)**
* unit tests for LifecyclePublisher and related classes

**main.ts (updated)**
* added dtslint LifecycleNode cases

**lifecycle-node-example.js (new)**
* demonstrates how to use LifecycleNode, register transition callbacks and change states.


Fix #589

wayneparrott committed:
Initial comment of lifecycle api, types, tests & example.

wayneparrott commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,372 @@
+  Nan::Set(named_srv_obj, Nan::New("name").ToLocalChecked(),
+           Nan::New(service_name).ToLocalChecked());
+
+  // Note: the following hack is to avoid a 'double free()' error that arises
+  // when overlapping RclHandles are created for a rcl_lifecycle_state_model and
+  // for each of the state-model's 4 services. The deleter for the state-model
+  // RclHandle calls rcl_lifecycle_state_machine_fini(state-model) which frees
+  // each of the state-model services. Then the RclHandle deleter for each
+  // state-model service is invoked resulting in a call to free() on the already
+  // freed service ptr.
+  // *** Boom - double free error ***.
+  //
+  // The work-around (hack) is to prevent the RclHandlers wrapping each of the
+  // state-machine services from calling free() on their service ptr. This is
+  // accomplished by setting each of the service RclHandlers' pointer field to
+  // nullptr causing their Reset() method to return immediately when invoked.
+  //
+  // The implementation of this approach replaces the nop deleter function on
+  // each service RclHandler with a new function that is passed a pointer to the
+  // service RclHandler. The deleter function then sets the service RclHandler
+  // parameter's pointer to nullptr. All joy - no double free()!
+  auto srv_handle =
+      RclHandle::NewInstance(service, nullptr, [] { return RCL_RET_OK; });
+  RclHandle* rclHandle = RclHandle::Unwrap<RclHandle>(srv_handle);
+  RclHandle::Unwrap<RclHandle>(srv_handle)->set_deleter([rclHandle] {
+      rclHandle->set_ptr(nullptr);
+      return RCL_RET_OK;
+    });
```
I would appreciate a close critique of this unit of code where I worked around an exception where free() was being called twice on the same underlying rcl unit of memory. In a nutshell, rcl provides a rcl_lifecycle_state_model that contains a state-model and also instances of Services for the lifecycle_msgs/srv/* interfaces. The LifecycleNode implementation accesses these Services from the rcl_lifecycle_state_model, creates RclHandles for each and then installs them in the node's services list. When the node is gc'ed the deleter for the RclHandle wrapping the rcl_lifecycle_state_model properly cleans up it's memory and releases the services it manages. The RclHandles wrapping each of the state_model services are calling free() on their service pointer which has already been freed by it's containing rcl_lifecycle_state_model. This is causing a segfault.

To avoid this 2nd free() I'm hacking each service RclHandler deleter such that it sets it's pointer to a service to null. Doing this prevents RclHandle#Reset() from calling free(). No more **double free()**.

Hope this makes sense.

minggangw replied:
This is a new scenario of `RclHandle`, currently, the pointer was freed by default when calling `Reset()`. I suggest we could extend the `RclHandle` class to make the code consistent

For the declaration:
```c
  // |handle| will be freed if |will_delete| is true.
  static v8::Local<v8::Object> NewInstance(
      void* handle, RclHandle* parent = nullptr,
      std::function<int()> deleter = [] { return 0; },
      will_delete = true)
```

For the implementation:
```c
if (will_delete_)
  free(pointer_);
```

than we could call like
```c
 auto srv_handle =
      RclHandle::NewInstance(service, nullptr, std::function<int()>(), false);
```

Also, I notice that we have https://github.com/RobotWebTools/rclnodejs/blob/711072d689acc7221cbb8e289ca552db2c423c4b/src/rcl_handle.cpp#L81-L86, not sure if we could leverage that.

koonpeng replied:
I suggest to instead make the `deleter`'s job to free the memory, the default delete will call free but you can pass in a custom deleter to perform other actions.

wayneparrott replied:
see issue #728

minggangw commented on index.js:
```diff
@@ -424,3 +429,30 @@ const TimeSource = require('./lib/time_source.js');
+  namespace = '',
+  context = Context.defaultContext(),
+  options = NodeOptions.defaultOptions,
+  nodeClazz
```
nodeClass? I googled `clazz` and found it was used in Java commonly :)

wayneparrott replied:
Renaming to `nodeClass`. From my zillion years developing with Java I used 'zz' without giving it real thought.

minggangw commented on lib/lifecycle.js:
```diff
@@ -0,0 +1,748 @@
+   * @returns {boolean} Return true if an error message has been assigned; otherwise return false.
+   */
+  hasErrorMsg() {
+    return this.isSuccess() ? false : !!this._errorMsg;
```
How about `return !this.isSuccess() && this._errorMsg;`

wayneparrott replied:
+1

minggangw commented on src/rcl_lifecycle_bindings.hpp:
```diff
@@ -0,0 +1,31 @@
+#define RCLNODEJS_RCL_LIFECYCLE_BINDINGS_HPP_
+
+#include <nan.h>
+#include <rcl/rcl.h>
```
these headers can be removed

minggangw commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,372 @@
+#include <rcl/error_handling.h>
+#include <rcl/rcl.h>
+
+#include <iostream>
```
I think we don't need this.

minggangw reviewed (commented):
Thanks for implementing this HUGE pull request and I look through it quickly, generally looks good! As I don't understand lifecycle comprehensively, I'd like to invite @koonpeng to review it, thanks!

koonpeng commented:
this is quite a large PR, I may need some time to review this :sweat_smile:. I don't have experience with lifecycle nodes as well so I will just try to review based on my knowledge of rcl.

wayneparrott commented:
@koonpeng @minggangw et al, please ping me if I can be of assistance during the review.

At work I usually start reviews by asking the dev how comfortable he/she is with a specific unit of code, are there any hot spots or overly complex areas that he has concerns. With that I'll share 2 areas that need discussion:

1. I've already highlighted that the implementation makes 4 Services created by the rcl lifecycle api visible to rclnodejs. A work-around of the RclHandler Reset behavior was implemented to avoid double free() error of the Service ptr.  

edited @wayneparrott - Please hold off on further review. While describing design issue #2 below I identified a simplification which I will submit an update later today.

~~2. The lifecycle.js module requires an `initialize()` function to be called before it is runnable. I'm not fully comfortable with this. This requirement arises because of a dependency cycle that arises because I was trying to use lifecycle state machine constants defined in the lifecycle [State and Transition messages](https://github.com/ros2/rcl_interfaces/tree/master/lifecycle_msgs) rather than hard-code those same constants in JS code.  An alternate approach is to hard-code the lifecycle state machine constants defined in the lifecyle State and Transition messages. **I'm not opposed to the latter approach.** But wanted to get the insight of others before doing so.~~

~~tldr;
Here's the details behind this and an alternate approach:~~

~~The `State` and `Transition` message definitions include the state id constants used by the lifecycle state machine. The implementation loads instances of these message types at runtime in order to reference their state machine constants. This dependency leads to the Lifecycle module not being runnable until message files have been generated which happens at rclnodejs installation and at `rlcnodejs.init()` time. Here's where it gets tricky, index.js loads the lifecycle module and exposes it through a new public lifecycle property, i.e., `rclnodejs.lifecycle`. To work around this chicken-egg dependency the lifecycle module has an initialize() function that must be called after calling rclnodejs.init(). At this time, lifecycle.initialize() is called implicitly as a part of rclnodejs.createLifecycleNode().~~

~~Summary of the dependency:  rclnodejs -> lifecycle module -> generated messages -> rclnodejs.init()~~

~~**The lifecycle module dependency upon the State and Transition messages could be eliminated by hard-coding their state machine constants in JS code rather than pulling them from these 2 messages. I was hoping to avoid hardcoding the constants. But as a consequence, the lifecycle module has to wait for rclnodejs.init() to complete before it can safely load the 2 messages.**~~

~~So before going too deep into the review, let's decide if the current design with dependencies on the message constants is acceptable or if we should work to remove the dependency on the State and Transition messages, i.e., hard code the  constants. **At this time I'm very open to hard coding the constants. Thoughts?**
tldr;~~

wayneparrott committed:
Rebase on 09766e3 and updates for preliminary lifecycle review.

Key Changes based on @minggangw comments and rethinking clunky
initialalization of StateInterface and TransitionInterface.
rcl_handle.hpp/cpp
* added `free_ptr_` bool property to RclHandle

lifecycle.js
* removed initialize() function and refactored to use lazy
initialization of StateInterface and TransitionInterface
* update shutdown() to use rcl_lifecycle_shutdown_label

wayneparrott commented:
The latest commit should be the one. It includes @minggangw earlier suggestions; especially see rclhandler.cpp free_ptr_. In lifecycle.js I switch to using lazy initialization of a couple of message interfaces which I feel is a good simplification from my initial commit.

koonpeng commented on src/rcl_handle.cpp:
```diff
@@ -86,7 +86,8 @@ NAN_METHOD(RclHandle::Dismiss) {
 
 v8::Local<v8::Object> RclHandle::NewInstance(void* handle, RclHandle* parent,
-                                             std::function<int()> deleter) {
+                                             std::function<int()> deleter,
```
I suggest to instead make the `deleter`'s job to free the memory, the default delete will call free but you can pass in a custom deleter to perform other actions. This follows closer to how `std::shared_ptr` and other v8 functions that takes in a custom deleter works.

wayneparrott replied:
@koonpeng @minggangw I can see the merit to your recommendation to delegate managing ptr memory to a deleter function. Yet I feel making such a change is beyond the scope of this PR due to the large impact it would have across the addon codebase. My preference is to continue with my original technique/hack, i.e., where my deleter function sets the rclHandle's ptr to null, and open a separate issue for how RclHandle handles free it's ptr.

minggangw replied:
+1 We'd better focus on the lifecycle feature within this PR, considering we have already had a lot of changes. @wayneparrott please go-ahead to open an issue, as @koonpeng said we should have a completely user-defined deleter.

wayneparrott replied:
Opened issue #728

koonpeng commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,363 @@
+
+static v8::Local<v8::Object> wrapState(const rcl_lifecycle_state_t* state) {
+  v8::Local<v8::Object> jsState = Nan::New<v8::Object>();
+  Nan::Set(jsState, Nan::New("id").ToLocalChecked(), Nan::New(state->id));
+  Nan::Set(jsState, Nan::New("label").ToLocalChecked(),
+           Nan::New(state->label).ToLocalChecked());
```
I don't know much about native c++ modules, would it be better if we use v8 accesors to bind these properties? https://v8.dev/docs/embed#accessing-dynamic-variables and https://github.com/nodejs/nan/blob/master/doc/methods.md#api_nan_set_accessor.

We might also want to make them readonly.

wayneparrott replied:
My initial thought regarding your suggestion to bind JS objects to C++ object wrappers on rcl C structs is a more object-oriented approach and a departure from the current functional style addon api design. I'm hesitant to embark on a new approach without broader discussion and agreement. It seems like a slippery slope that could quickly get very complicated when generally applied.

I realize my impl dynamically creates JS data-transfer objects and copies data into them rather than directly referencing the C data from JS objects. In this case is seemed like an OK tradeoff of a little extra memory per state/transition data for a simple implementation. Maybe this is the wrong attitude given this code could be running on a smaller system, e.g., raspi. 

My preference is to continue with the current api design. But I'm open (team player) to a general change towards an object-oriented wrapping approach. Before we go there, consider the potential implementation complexity over our functional model. My expectation is adopting an OO approach will result in creating parallel C++ hierarchies wrapping rcl C structs and api with corresponding JS bindings. So in the lifecycle case that could involve creating C++ representations for Lifecycle StateMachine, State, Transition, TransitionDescription and Service.

@minggangw @koonpeng  Thoughts?

koonpeng replied:
I don't mind the current design, the only small problem I see is that some functions are creating a temporary object which are immediately discarded on the js side. But I don't think these functions are called often enough that it matters

Currently these are only simple plain data objects, but if we want to add methods to them in the future, it could create quite a lot of extra complexity that might not be worth the trouble.

wayneparrott replied:
Agree. Thx for the feedback.

koonpeng commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,363 @@
+  Nan::Set(jsTransition, Nan::New("id").ToLocalChecked(),
+           Nan::New(transition->id));
+  Nan::Set(jsTransition, Nan::New("label").ToLocalChecked(),
+           Nan::New(transition->label).ToLocalChecked());
```
See above comment in `SetAccessor`.

wayneparrott replied:
see comments above.

koonpeng commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,363 @@
+  std::string service_name;
+  rcl_service_t* service;
+  if (lifecycle_srv_field_name.compare("srv_get_state") == 0) {
+    service_name = "~/get_state";
```
Should we be using `rcl_service_get_service_name` instead of hardcoding this?

wayneparrott replied:
+1, I completely missed this api and hardcoded as my fallback plan. Will make this change.

koonpeng commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,363 @@
+  } else if (lifecycle_srv_field_name.compare("srv_change_state") == 0) {
+    service_name = "~/change_state";
+    service = &state_machine->com_interface.srv_change_state;
+  }
```
missing find `else` so we may end up creating a `RclHandle` to a random access or nullptr (does c++ initialize pointers?).

wayneparrott replied:
+1. Added the following snippet:
```
THROW_ERROR_IF_EQUAL(
                   nullptr,
                   service,
                   "Service not found.");
```

koonpeng commented on src/rcl_lifecycle_bindings.cpp:
```diff
@@ -0,0 +1,363 @@
+  rcl_ret_t ret = rcl_lifecycle_trigger_transition_by_label(
+      state_machine, transition_label.c_str(), publish);
+
+  if (ret == RCL_RET_OK) {
```
throw on error?

wayneparrott replied:
+1. Using this macro now throughout the file.
```
  THROW_ERROR_IF_NOT_EQUAL(RCL_RET_OK,
                           rcl_lifecycle_trigger_transition_by_id(
                              state_machine, transition_id, publish),
                           rcl_get_error_string().str);
```

koonpeng commented on lib/lifecycle.js:
```diff
@@ -0,0 +1,728 @@
+/**
+ * The state of the lifecycle state model.
+ */
+class State {
```
We are already creating a new object with `wrapState` in the c++ side, we could avoid creating redundant objects by either using the object from c++, or bind the values on c++ side and create a copy of it here to avoid unintended modifications by the user.

wayneparrott replied:
See previous comments.

koonpeng commented on lib/lifecycle.js:
```diff
@@ -0,0 +1,728 @@
+   * Create an object representation of state properties.
+   * @return {SerializedState} The object.
+   */
+  asMessage() {
```
Doesn't this return the same thing except it removes the prototype chain?

wayneparrott replied:
+1. Removed asMessage() methods from State, Transition and TransitionDescription.

koonpeng commented on lib/lifecycle.js:
```diff
@@ -0,0 +1,728 @@
+   * 
+   * @returns {object} The object representation.
+   */
+  asMessage() {
```
Doesn't this return the same thing except it removes the prototype chain?

wayneparrott replied:
+1. see above.

koonpeng reviewed (commented):
Echoing back to `SetAccessors`, I wonder if it's possible to bind all the rcl structs to js objects.

wayneparrott committed:
Fix lint issue.

minggangw reviewed (approved):
I have no further comments, lgtm, and thanks for implementing this important feature, great progress!

minggangw commented:
I think we could move forward, @koonpeng any comments? @wayneparrott please merge it when it's good to go (soft reminder: please select "Squash and merge" option)

koonpeng reviewed (approved):
LGTM!
